synapse/storage/data_stores/main/events.py

```diff
@@ -1957,6 +1973,105 @@ def insert_labels_for_event_txn(
             ],
         )
 
+    def insert_event_expiry_txn(self, txn, event_id, expiry_ts):
+        """Save the expiry timestamp associated with a given event ID.
+
+        Args:
+            txn (LoggingTransaction): The database transaction to use.
+            event_id (str): The event ID the expiry timestamp is associated with.
+            expiry_ts (int): The timestamp at which to expire (delete) the event.
+        """
+        return self._simple_insert_txn(
+            txn=txn,
+            table="event_expiry",
+            values={"event_id": event_id, "expiry_ts": expiry_ts},
+        )
+
+    def expire_event(self, event):
+        """Redact an event that has expired, and delete its associated expiry timestamp.
+
+        Args:
+             event (events.EventBase): The event to delete.
+        """
+        # Prune the event's dict then convert it to JSON.
+        pruned_json = encode_json(prune_event_dict(event.get_dict()))
+
+        def delete_expired_event_txn(txn):
+            # Update the event_json table to replace the event's JSON with the pruned
+            # JSON.
+            self._censor_event_txn(txn, event.event_id, pruned_json)
+
+            # Delete the expiry timestamp associated with this event from the database.
+            self._simple_delete_txn(
+                txn, table="event_expiry", keyvalues={"event_id": event.event_id}
+            )
+
+            # We need to invalidate the event cache entry for this event because we
+            # changed its content in the database.
+            txn.call_after(self._get_event_cache.invalidate, (event.event_id,))
+            # Send that invalidation to replication so that other workers also invalidate
+            # the event cache.
+            self._send_invalidation_to_replication(
```

Comment:
`_invalidate_cache_and_stream` is a thing

Reply:
Which is what I'd normally use, but it doesn't work with the `get_event` cache since its type is `Cache` not `CacheDescriptor` ans therefore has no method/attribute `__name__`, which `_invalidate_cache_and_stream` requires.

Reply:
well ok, a comment would be useful then.

Reply:
Yep, agreed, done in https://github.com/matrix-org/synapse/pull/6409/commits/07678989be92ee9b233d488e3dd83670f3bdf01f